src/binaryutils.cpp

```diff
@@ -0,0 +1,259 @@
+
+#include <inference_engine.hpp>
+
+namespace ovms {
+
+int getMatTypeFromTensorPrecision(InferenceEngine::Precision tensorPrecision) {
+    switch (tensorPrecision) {
+    case InferenceEngine::Precision::FP32:
+        return CV_32F;
+    case InferenceEngine::Precision::FP16:
+        return CV_16F;
+    case InferenceEngine::Precision::I16:
+        return CV_16S;
+    case InferenceEngine::Precision::U8:
+        return CV_8U;
+    case InferenceEngine::Precision::I8:
+        return CV_8S;
+    case InferenceEngine::Precision::U16:
+        return CV_16U;
+    case InferenceEngine::Precision::I32:
+        return CV_32S;
+    default:
+        return -1;
+    }
+}
+
+bool isPrecisionEqual(int matPrecision, InferenceEngine::Precision tensorPrecision) {
+    int convertedTensorPrecision = getMatTypeFromTensorPrecision(tensorPrecision);
+    if (convertedTensorPrecision == matPrecision) {
+        return true;
+    }
+    return false;
+}
+
+template <typename T>
+std::vector<T> reorder_to_nchw(const T* nhwcVector, int rows, int cols, int channels) {
+    std::vector<T> nchwVector(rows * cols * channels);
+    for (int y = 0; y < rows; ++y) {
+        for (int x = 0; x < cols; ++x) {
+            for (int c = 0; c < channels; ++c) {
+                nchwVector[c * (rows * cols) + y * cols + x] = reinterpret_cast<const T*>(nhwcVector)[y * channels * cols + x * channels + c];
+            }
+        }
+    }
+    return std::move(nchwVector);
+}
+
+cv::Mat convertStringValToMat(const std::string& stringVal) {
+    std::vector<unsigned char> vectordata(stringVal.begin(), stringVal.end());
+    cv::Mat data_mat(vectordata, true);
+
+    return cv::imdecode(data_mat, cv::IMREAD_UNCHANGED);
+}
+
+Status convertPrecision(const cv::Mat& src, cv::Mat& dst, const InferenceEngine::Precision requestedPrecision) {
+    int type = getMatTypeFromTensorPrecision(requestedPrecision);
+    if (type == -1) {
+        return StatusCode::INVALID_PRECISION;
+    }
+
+    src.convertTo(dst, type);
+    return StatusCode::OK;
+}
+
+Status resizeMat(const cv::Mat& src, cv::Mat& dst, const std::shared_ptr<TensorInfo>& tensorInfo) {
+    if(tensorInfo->getLayout() == InferenceEngine::Layout::NCHW)
+    {
+        int cols = tensorInfo->getShape()[3];
+        int rows = tensorInfo->getShape()[2];
+        cv::resize(src, dst, cv::Size(cols,rows));
+        return StatusCode::OK;
+    }
+    else if (tensorInfo->getLayout() == InferenceEngine::Layout::NHWC)
+    {
+        int cols = tensorInfo->getShape()[2];
+        int rows = tensorInfo->getShape()[1];
+        cv::resize(src, dst, cv::Size(cols,rows));
+        return StatusCode::OK;
+    }
+
+    return StatusCode::UNSUPPORTED_LAYOUT;
+}
+
+Status validateNumberOfChannels(const std::shared_ptr<TensorInfo>& tensorInfo,
+    const cv::Mat input) {
+    // Network and input must have the same number of shape dimensions. 
+    if ((unsigned int)(input.channels()) != tensorInfo->getShape()[1]) {
```

Comment:
Here number of color channels does not need to be at index 1. Maybe introduce method in TensorInfo `getColorChannels()`? Depending on layout it would return correct number of channels. Or maybe OpenVINO has utility functions for this?

Reply:
You also do not check if tensor has at least 2 elements.